### PR TITLE
Add a junit5 extension

### PR DESCRIPTION
**Pull request by tombentley**
This PR does  one significant thing

* Add support for a Junit5 extension that allows declarative test clusters via annotation

It also does  a few minor things:

* Add method `getClusterId` to `KafkaCluster`, which can be useful in certain contexts.
* Add method `getKafkaClientConfiguration(String user, String password)` to `KafkaCluster`. This allows for more convenient testing with multiple SASL users when those used are declared by annotation, including writing tests for bad credentials/unknown users against test `KafkaClusters` with some users defined.
* Moves the package for the test infrastructure, because it doesn't depend on the proxy work in any way.

Examples for how to declare test clusters are in the test for the extension. The API is such as it is to minimise the need for changes in future when support for ZK mode gets removed from Kafka. E.g. the `@ZooKeeperCluster` annotation can be deprecated and removed without affecting tests which don't explicitly depend on ZK. 

I'm creating a _draft_ PR because I don't actually want to merge this in its current form. I think this is useful for other projects (in a similar way to the Debezium Kafka cluster test code). So I'd like to factor both the junit extension and the `KafkaCluster` infrastructure into its own distinct repo under the kroxylicious org. That would allow us to do a release of this test infra, allowing other projects to use it while we work towards a release of kproxy.

### CODE REVIEW

File: integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterConfig.java [hunk 1]
```diff
@@ -152,18 +152,34 @@ public Stream<ConfigHolder> getBrokerConfigs(Supplier<KafkaEndpoints> endPointCo
...
             kafkaConfig.put(SaslConfigs.SASL_MECHANISM, saslMechanism);
 
-            Map<String, String> users = getUsers();
-            if ("PLAIN".equals(saslMechanism) && !users.isEmpty()) {
-                var first = users.entrySet().iterator().next();
-                kafkaConfig.put(SaslConfigs.SASL_JAAS_CONFIG,
-                        String.format("org.apache.kafka.common.security.plain.PlainLoginModule required username=\"%s\" password=\"%s\";",
-                                first.getKey(), first.getValue()));
+            if ("PLAIN".equals(saslMechanism)) {
+                if (user != null && password != null) {
```
Comment (SamBarker) [inline integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterConfig.java:178]: Does enforcing not null on both exclude som edge case testing?

File: integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterExecutionMode.java [hunk 2]
```diff
@@ -4,7 +4,7 @@
  * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-package io.kroxylicious.proxy.testkafkacluster;
+package io.kroxylicious.test.kafkacluster;
```
Comment (SamBarker) [inline integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterExecutionMode.java:7]: I wonder if the package should be:

```suggestion
package io.kroxylicious.testSupport.kafkacluster;
```

just to make it explicit that it is "production code"

File: integrationtests/src/test/java/io/kroxylicious/test/kafkacluster/extension/KafkaClusterExtensionTest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+import io.kroxylicious.test.kafkacluster.ContainerBasedKafkaCluster;
+import io.kroxylicious.test.kafkacluster.KafkaCluster;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@ExtendWith(KafkaClusterExtension.class)
```
Comment (SamBarker) [inline integrationtests/src/test/java/io/kroxylicious/test/kafkacluster/extension/KafkaClusterExtensionTest.java:26]: This is possibly a use case for nested test cases. So that the unit test for the extension doesn't depend on the extension itself. I'd almost make the argument that this is an "integration" test.

File: integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/extension/BrokerCluster.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * @ExtendWith(KafkaClusterExtension.class)
+ * public class MyTest {
+ *     @Test
+ *     public void testKafkaClusterParameter(
+ *             @BrokerCluster(numBrokers = 3)
+ *             @KRaftCluster(numControllers = 3)
+ *             @SaslPlainAuth({
+ *                 @UserPassword(user="alice", password="foo"),
+ *                 @UserPassword(user="bob", password="bar")
+ *             })
```
Comment (SamBarker) [inline integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/extension/BrokerCluster.java:65]: The annotation explosion was one concern I had when suggesting the extension originally. 

Did you experiment with something like
`@KafkaCluser(clusterConfig=[])` 

Then pass the `KafkaClusterConfig.builder()` into each clusterConfig instances i.e. 

```java
public interface ClusterConfig {

KafkaClusterConfig.builder customise(KafkaClusterConfig.builder builder);
}
```

So the BrokerCluster would implement as follows:
```java

public KafkaClusterConfig.builder customise(KafkaClusterConfig.builder builder) {                 
     return builder.brokersNum(this.numBrokers());
}
```

**Review by SamBarker — APPROVED** [review]
Awesome to see this @tombentley! 

I have a few thoughts about the mechanism for handling the cluster config but I think this could merge as a useful increment as is and revisit the config separately.

File: integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/extension/BrokerCluster.java (hunk 4 above)
Comment (tombentley) [reply integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/extension/BrokerCluster.java:65]: I was playing more with this code on Friday and thinking about this over the weekend... One direction we could go for this whole idea would be as follows...

Support more provisioning strategies for test clusters. I.e. currently we have in-VM (which is fast, but not terribly flexible since you can only easily spin up a single version of the broker), and in-container (which is very flexible but slow due to testcontainers – even though we're using a native broker). There is scope for others: 
* For example you could download and run a native broker in a forked process which would allow testing different broker versions but still wouldn't be _as customisable_ as as using containers. E.g. you couldn't do a good job with SSL because you can't mess with hostnames/SAN certificates like you could in containers. 
* You could provision a cluster on a KaaS, or use an already-provisioned cluster on a KaaS.

The extension allows you to decouple your test code (which needs a cluster, with some configuration that might be specific to the test) from how it's provisioned. Abstractly, the annotations provide constraints on which provisioning mechanisms can be used. Within those constraints the extension is free to pick "the best" provisioning mechanism. Constraints could also be added externally to the code. This would allow you to default to testing locally using the fastest mechanism that satisfied the constraints (saving developer time), but configure a CI pipeline to test on (something closer to) the target platform (by adding extra constraints).

Looked at in this light, then, the annotation(s) are one way of expressing such constraints. Here are some constraints we might reasonably want to support:

* I want a KRaft-based/ZK-based cluster specifically
* I want at least N brokers
* I want _this particular broker version_ (or for a parameterised/templated test, _this range of broker versions_)
* I want a cluster with _this specific user_ preconfigured (using whatever SASL mechanism, I don't care)
* I want a cluster with _this specific user and password_ configured, using _this specific SASL_ mechanism.

Logically there are plenty of others too (e.g. the extension could provision things within the cluster as well as the cluster itself, such as quotas, ACLs). 

If this degree of abstraction is valuable we'd need:

1. To make cluster provisioning strategies pluggable (e.g. using `ServiceLoader`)
2. Make the constraining annotations flexible enough to cope with our needs (and other people's needs). Ideally they might even be extensible (e.g. using a meta-annotation on some custom annotation honoured by a pluggable provisioning strategy).

I'm not saying we should do this all at once and right now. But the decisions we make here could enable/prevent doing this in the future. My gut feeling is that if we did a good job with this, documented it well and promoted it then this extension could be broadly useful for many people in Red Hat and beyond. Or maybe I'm massively overengineering this. Wdyt?

File: integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterExecutionMode.java (hunk 2 above)
Comment (tombentley) [reply integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterExecutionMode.java:7]: Perhaps we go so far as to put the `KafkaCluster` code into `io.kroxy.cluster` and the extension into `io.kroxy.junit5` as separate modules in a new repo. 

Also: Package names are supposed to be strictly lower case for deeply historical reasons.

Comment (SamBarker) [reply integrationtests/src/main/java/io/kroxylicious/test/kafkacluster/KafkaClusterExecutionMode.java:7]: both your suggested package names are a lot better than `testSupport` 😀

**Comment by k-wall** [comment]
I believe this PR is now dead, closing.
